### PR TITLE
Feature/parameter requirements

### PR DESCRIPTION
**Pull request by wandmagic**
# Committer Notes

introduce
- has-required-parameter constraint

I'd like to also make sure that choice type constraints are one of the values but, i'll save that for another PR this one is heavy enough.

### All Submissions:

- [x] Have you selected the correct base branch per [Contributing](https://github.com/GSA/fedramp-automation/blob/master/CONTRIBUTING.md) guidance?
- [x] Have you set "[Allow edits and access to secrets by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/GSA/fedramp-automation/pulls) for the same update/change?
- [x] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] If applicable, have all [FedRAMP Documents Related to OSCAL Adoption](https://github.com/GSA/fedramp-automation) affected by the changes in this issue have been updated.?
- [x] If applicable, does this PR reference the issue it addresses and explain how it addresses the issue?

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

### CODE REVIEW

**Comment by wandmagic** [comment]
addresses [#1006](https://github.com/GSA/fedramp-automation/issues/1006)

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 1]
```diff
@@ -62,6 +62,29 @@
...
+            <expect id="component-has-authentication-method" target="//component[(@type='system' and prop[@name='leveraged-authorization-uuid']) or (@type='interconnection') or (@type=('service', 'software') and not(prop[@name='leveraged-authorization-uuid']) and prop[@name='implementation-point' and @value='external']) or (@type=('service', 'software') and prop[@name='implementation-point' and @value='internal'] and prop[@name='communicates-externally' and @value='yes' and @ns='https://fedramp.gov/ns/oscal'])]" test="count(prop[@ns='https://fedramp.gov/ns/oscal' and @name='authentication-method']) >= 1" level="ERROR">
+                <formal-name>Component Has Authentication Method</formal-name>
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/4-ssp-template-to-oscal-mapping/#leveraged-fedramp-authorized-services"/>
+                <message>A FedRAMP SSP MUST include at least one authentication method for each leveraged system.</message>
+            </expect>
```
Comment (aj-stein-gsa) [inline src/validations/constraints/fedramp-external-constraints.xml:71]: Blocking: the rest looks solid, but was this inclusion intentional?

**Review by DimitriZhurkin — DISMISSED** [review]
Well done.

**Comment by wandmagic** [comment]
Thanks for the update, I wasn't aware of the aggregate params situation

**Review by aj-stein-gsa — CHANGES_REQUESTED** [review]
As requested, I reviewed the option's per @Rene2mt's review and I would like us to proceed with filtering on ODP parameters, and not require definiton of the classic (non-ODP parameters). More details in https://github.com/GSA/fedramp-automation/issues/1006#issuecomment-2550531101. Thanks for your patience in me sorting this decision out.

**Comment by wandmagic** [comment]
> As requested, I reviewed the option's per @Rene2mt's review and I would like us to proceed with filtering on ODP parameters, and not require definiton of the classic (non-ODP parameters). More details in [#1006 (comment)](https://github.com/GSA/fedramp-automation/issues/1006#issuecomment-2550531101). Thanks for your patience in me sorting this decision out.

so to get this correct, we need only params that include the string "odp" ??

**Comment by aj-stein-gsa** [comment]
> so to get this correct, we need only params that include the string "odp" ??

Correct, @wandmagic.

**Comment by Rene2mt** [comment]
The following metapath should grab all params excluding ones that aggregate `//param[not(prop[@name="aggregates"])]`

**Comment by wandmagic** [comment]
we now have a warning advising against the usage of aggregate params (instead of expressly forbidding them) and a constraint to make sure we have non-aggregate params

**Comment by aj-stein-gsa** [comment]
> we now have a warning advising against the usage of aggregate params (instead of expressly forbidding them) and a constraint to make sure we have non-aggregate params

Sorry for the delayed reply. I know we discussed this earlier in chat, but wanted to make it explicit here: I would recommend a constraint with `level="WARNING"` for aggregate params (i.e. those that are `odp` within their `@id`s).

**Comment by wandmagic** [comment]
> > we now have a warning advising against the usage of aggregate params (instead of expressly forbidding them) and a constraint to make sure we have non-aggregate params
> 
> 
> 
> Sorry for the delayed reply. I know we discussed this earlier in chat, but wanted to make it explicit here: I would recommend a constraint with `level="WARNING"` for aggregate params (i.e. those that are `odp` within their `@id`s).

Yes that is what is in the pr.

**Review by DimitriZhurkin — DISMISSED** [review]

**Review by DimitriZhurkin — APPROVED** [review]

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 2]
```diff
@@ -62,6 +62,30 @@
...
+                <formal-name>Required Parameters Must Be Set</formal-name>
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/6-security-controls/#responsible-roles-and-parameter-assignments"/>
+                <key-field target="@id"/>
+                <message>All parameters defined in the baseline MUST have corresponding set-parameter values in the SSP. Missing parameter: ({@id}).</message>
+            </index-has-key>
+            <expect id="aggregate-parameters-warning" target="//set-parameter" test="every $param in $aggregate-parameters satisfies @param-id != $param" level="WARNING">
+                <formal-name>No Aggregate Parameters in SSP</formal-name>
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/6-security-controls/#responsible-roles-and-parameter-assignments"/>
+                <message>A FedRAMP SSP MUST NOT set aggregate parameters directly. Parameter {@param-id} is an aggregate parameter that should not be set in the SSP.</message>
```
Comment (kyhu65867) [inline src/validations/constraints/fedramp-external-constraints.xml:87]: Must implies error as opposed to warning. Change must to should, like this: 
```suggestion
                <message>A FedRAMP SSP SHOULD NOT set aggregate parameters directly. Parameter {@param-id} is an aggregate parameter that should not be set in the SSP.</message>
```

**Review by kyhu65867 — CHANGES_REQUESTED** [review]

**Review by kyhu65867 — APPROVED** [review]

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 3]
```diff
@@ -62,6 +62,30 @@
             <let var="resolved-profile" expression="if ($profile-is-available) then resolve-profile(doc(resolve-uri($resolved-import-profile-href)))//catalog else //catalog"/>
             <let var="imported-controls-map" expression="map:merge($resolved-profile//control ! map:entry(@id,.))?*"/>
             <let var="implemented-requirements-map" expression="map:merge(//implemented-requirement ! map:entry(@control-id,.))?*"/>          
+            <let var="required-parameters-map" expression="map:merge($resolved-profile//param[not(prop[@name='aggregates'])] ! map:entry(@id,.))?*"/>          
+            <let var="aggregate-parameters" expression="$resolved-profile//param[prop[@name='aggregates']]/@id"/>
+            <let var="implemented-parameters-map" expression="map:merge(//set-parameter ! map:entry(@param-id,.))?*"/>          
+            <index name="index-implemented-parameters" target="$implemented-parameters-map">
+                <formal-name>Parameters Implemented in SSP</formal-name>
+                <description>All parameters defined in the SSP</description>
```
Comment (aj-stein-gsa) [inline src/validations/constraints/fedramp-external-constraints.xml:70]: Can we be a little more descriptive here for the sake of downstream users?


```suggestion
                <description>This constraint indexes all parameters in a FedRAMP SSP.</description>
```

**Review by DimitriZhurkin — APPROVED** [review]

**Review by Gabeblis — APPROVED** [review]

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 4]
```diff
@@ -62,6 +62,30 @@
...
+            <index name="index-required-parameters" target="$required-parameters-map">
+                <formal-name>Parameters from Profile</formal-name>
+                <description>All parameters defined in the resolved profile catalog</description>
+                <key-field target="@id"/>
+            </index>
+            <index-has-key id="has-required-parameters" target="$required-parameters-map" name="index-implemented-parameters" level="ERROR">
+                <formal-name>Required Parameters Must Be Set</formal-name>
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/6-security-controls/#responsible-roles-and-parameter-assignments"/>
+                <key-field target="@id"/>
+                <message>All parameters defined in the baseline MUST have corresponding set-parameter values in the SSP. Missing parameter: ({@id}).</message>
```
Comment (aj-stein-gsa) [inline src/validations/constraints/fedramp-external-constraints.xml:82]: Minor on active voice and style requirements.

```suggestion
                <message>A FedRAMP SSP must define all parameters for all controls from the imported baseline. The following parameters are defined in the baseline, but not properly set in the SSP: ({@id}).</message>
```

**Review by aj-stein-gsa — COMMENTED** [review]
Too little too, this and more in another PR.

File: src/validations/constraints/fedramp-external-constraints.xml (hunk 3 above)
Comment (wandmagic) [reply src/validations/constraints/fedramp-external-constraints.xml:70]: right on i'll include this in the next PR

File: src/validations/constraints/fedramp-external-constraints.xml (hunk 4 above)
Comment (wandmagic) [reply src/validations/constraints/fedramp-external-constraints.xml:82]: ditto

Comment (aj-stein-gsa) [reply src/validations/constraints/fedramp-external-constraints.xml:82]: Hold up on that there are a few others I'll just send it to you somewhere, no big deal.

**Pull request merged**